### DISCUSSION TITLE
Layer violation: domain classes should not depend on testing classes whatsoever

### DISCUSSION THREAD

**Discussion by matheusgontijo** [shopware/platform#3155] [RFC / Proposal]
**Effort**: medium

### Description
Oh, thank you so much for your williness to start this proposal channel. Highly appreciate the openness to listen to suggestions.

I usually do something very very particular during development. I usually get rid of the `vendor/shopware/****/Test|Tests|test|tests|...` directories from the projects I work on locally. Why? Because of the fact I hate when I search for a specific class on PhpStorm and oh my gooooodness, it seems there are thousands of thousands of the same files but testing-related. My searches become super ineffective since I open 10 files and all of them are **not** the ones I'm looking for. They are testing files. After opening 10 incorrect files, I eventually find the one I was looking for. That drives me crazy. Anyways, removing test directories is something I personally do, nothing too important. Just a little of context, so you are able to understand where I'm coming from.

The thing is that, by doing this process (removing test directories) I found out an inconsistency. I found that that Shopware "domain" classes rely on Shopware "testing" classes. In other words, classes from the domain layer are depending on classes from another layer (testing layer, let's call it). Classes from the domain layer should not, by any chance, depend on classes from testing layer.

Let's take the `Shopware\Core\Defaults` for example.

`Shopware\Core\Defaults` is relying on `Shopware\Core\Test\TestDefaults` and it should not.

```
<?php declare(strict_types=1);

namespace Shopware\Core;

use Shopware\Core\Test\TestDefaults;  // <<<<<<<<<<<<<<<<< HERE
```

There are basically two references to `TestDefaults`.

```
public const SALES_CHANNEL = TestDefaults::SALES_CHANNEL;
/** .... */
public const FALLBACK_CUSTOMER_GROUP = TestDefaults::FALLBACK_CUSTOMER_GROUP;
```

Maybe there is something I might not be considering BUT `TestDefaults` could be depending on `Defaults`, not the other way around. These data could be duplicated, but please let's not make `Defaults` depending on `TestDefaults`.

I would be fine with hard-coded duplicated values on this scenario, especially because we are talking about just two values, not a multitude of them. That should be a problem update values on both files. Better than having a testing dependency on the domain like that.

```
public const SALES_CHANNEL = '98432def39fc4624b33213a56b8c944d';
/** .... */
public const FALLBACK_CUSTOMER_GROUP = 'cfbd5018d38d41d8adca10d94fc8bdd6';
```

Again, it might be something I'm not seeing or not considering, but apparently it's doesn't seem correct this type of dependency.

After removing the test directories locally, I found out some other references on PHP classes:

```
grep -P '^use Shopware.*\\Test' . -R
``` 

```
./src/Core/Defaults.php:use Shopware\Core\Test\TestDefaults;
./src/Core/Framework/Demodata/Generator/CustomerGenerator.php:use Shopware\Core\Test\TestDefaults;
./src/Core/Framework/Demodata/Generator/RuleGenerator.php:use Shopware\Core\Test\TestDefaults;
./src/Core/Framework/Framework.php:use Shopware\Core\Framework\Test\DependencyInjection\CompilerPass\ContainerVisibilityCompilerPass;
./src/Core/Framework/Framework.php:use Shopware\Core\Framework\Test\RateLimiter\DisableRateLimiterCompilerPass;
./src/Core/Framework/RateLimiter/NoLimitRateLimiterFactory.php:use Shopware\Core\Framework\Test\RateLimiter\DisableRateLimiterCompilerPass;
./src/Core/TestBootstrapper.php:use Shopware\Core\Framework\Test\TestCaseBase\KernelLifecycleManager;
```

I haven't thought about all scenarios or all possible collateral effects, but I would like to let you know about this reverse dependency. Hope that helps. I'm more than open for conversations. Let me know what you think about it.

Once again, thank you so much for your hard work and willingness hear the community.

### Benefits

Layer separation. Domain classes should not depend on testing classes whatsoever.

**Comment by keulinho** [comment]
Hey @matheusgontijo,

thanks for your proposal! In general i completely agree with you, especially as we already thought about shipping the packages over composer without the `/tests` folders.

But i think this proposal does not necessarily have to be addressed for the next major version, but can be fixed in a backwards compatible manner. 

Let's take a look on how each of the dependencies can be removed without introducing a break.

### Defaults.php

As @shyim  already mentioned on twitter the uses of the dependency are already deprecated and will be removed with the next major, but instead we could duplicate the hard coded defaults for the time being, as you already proposed.

### DemodataGenerators

Both demodata generators depend on the default customer group, but instead of accessing the const of the `TestDefaults` we could fetch any customer group where the `displayGross` property is true and use that value instead.

### Framework.php

In this case the uses are totaly valid IMHO, as the mentioned CompilerPasses will only be registered if the `APP_ENV = 'test'` and just importing the classes with a `use` statement will not break if the classes are not present.

### NoRateLimiterFactory

That class is only used inside the tests, so IMHO we should copy the class into the test folder and always use the one from the test folder and deprecate the old one. To already break the dependency we can remove the offending `if` clause completely and of course mark the whole class as deprecated as it should only be used inside tests.

### TestBootstrapper

IMHO the usage in the TestBootstrapper is also fine, as the class name already says that class is used to bootstrap the application for the test execution, if you remove the tests folder you won't execute the tests and therefore the class will never be executed, so it should not break anything. I'm not quite sure if that class shouldn't be moved to the tests folder too :thinking: 

## How to go forward?

The fastest way would be to provide a PR to fix one or all of the violations as i laid out before, there is no need to delay the fix for the next major. :wink:

**Comment by JoshuaBehrens** [comment]
So as class map excludes don't stop composer to load these files (could it be that the pattern is wrong? `*/Test` should be `**/Test` right?) should this be tested with the classmap-authoritative flag? This way it should break hard and you can instantly see the results of the separation of tests and production code

**Comment by keulinho** [comment]
We'll close this proposal for now, if that issue still exists feel free to reopen it as an issue.
